civis/ml/_helper.py

```diff
@@ -58,3 +62,261 @@ def list_models(job_type="train", author=SENTINEL, client=None, **kwargs):
+    readers : dict::
+        - users : list::
+            - id : integer
+            - name : string
+        - groups : list::
+            - id : integer
+            - name : string
+    writers : dict::
+        - users : list::
+            - id : integer
+            - name : string
+        - groups : list::
+            - id : integer
+            - name : string
+    owners : dict::
+        - users : list::
+            - id : integer
+            - name : string
+        - groups : list::
+            - id : integer
+            - name : string
+    total_user_shares : integer
+        For owners, the number of total users shared. For writers and readers,
+        the number of visible users shared.
+    total_group_shares : integer
+        For owners, the number of total groups shared. For writers and readers,
+        the number of visible groups shared.
+    """
+    kwargs = {}
+    if send_shared_email != 'DEFAULT':
+        kwargs['send_shared_email'] = send_shared_email
+    if share_email_body != 'DEFAULT':
+        kwargs['share_email_body'] = share_email_body
+    return _share_model(id, group_ids, permission_level, entity_type='groups',
+                        client=client, **kwargs)
+
+
+def _share_model(job_id, entity_ids, permission_level, entity_type,
+                 client=None, **kwargs):
+    """Share a container job and all run outputs with requested entities"""
+    client = client or APIClient()
+    if entity_type not in ['groups', 'users']:
+        raise ValueError("'entity_type' must be one of ['groups', 'users']. "
+                         "Got '{0}'.".format(entity_type))
+
+    log.debug("Sharing object %d with %s %s at permission level %s.",
+              job_id, entity_type, entity_ids, permission_level)
+    _func = getattr(client.scripts, "put_containers_shares_" + entity_type)
+    result = _func(job_id, entity_ids, permission_level, **kwargs)
+
+    # CivisML relies on several run outputs attached to each model run.
+    # Go through and share all outputs on each run.
+    runs = client.scripts.list_containers_runs(job_id, iterator=True)
+    for run in runs:
+        log.debug("Sharing outputs on %d, run %s.", job_id, run.id)
+        outputs = client.scripts.list_containers_runs_outputs(job_id, run.id)
+        for _output in outputs:
+            if _output['object_type'] == 'File':
+                _func = getattr(client.files, "put_shares_" + entity_type)
+                if _output['name'] == 'log.txt' and permission_level == 'read':
+                    # Require "write" level permission or higher to view
+                    # the debug logs.
+                    continue
+                obj_permission = permission_level
+            elif _output['object_type'] == 'Project':
+                _func = getattr(client.projects, "put_shares_" + entity_type)
+                if permission_level == 'read':
+                    # Users must be able to add to projects to use the model
```

Comment:
Just for my own understanding, why is this?

Reply:
Hmm, I didn't think about whether a "read" user should be able to score based off the original model rather than a clone. That reasoning makes sense to me, thanks.

Follow-up question: what permissions does the original user have on a scoring job created by the new user in this way?

Reply:
The original user wouldn't have any special permissions on the scoring job by default. If you don't have "read" permission on the scoring job, it wouldn't be visible to you in the project. I think this is expected behavior -- I could imagine the "scoring" project being filled with scoring jobs, and no one user being able to see all of them.

Reply:
Sounds good. I agree that this is expected behavior.